mesonbuild/wrap/wrap.py

```diff
@@ -226,27 +226,56 @@ def get_file(self):
+            subprocess.check_call(['git', 'remote', 'add', 'origin', self.wrap.get('url')],
                                   cwd=self.dirname)
-        push_url = self.wrap.values.get('push-url')
-        if push_url:
-            subprocess.check_call(['git', 'remote', 'set-url',
-                                   '--push', 'origin', push_url],
+            revno = self.wrap.get('revision')
+            subprocess.check_call(['git', 'fetch', '--depth', self.wrap.values.get('depth'), 'origin', revno],
                                   cwd=self.dirname)
+            subprocess.check_call(['git', 'checkout', revno], cwd=self.dirname)
+            if self.wrap.values.get('clone-recursive', '').lower() == 'true':
+                subprocess.check_call(['git', 'submodule', 'update',
+                                       '--init', '--checkout', '--recursive', '--depth', self.wrap.values.get('depth')],
+                                      cwd=self.dirname)
+            push_url = self.wrap.values.get('push-url')
+            if push_url:
+                subprocess.check_call(['git', 'remote', 'set-url',
+                                       '--push', 'origin', push_url],
+                                      cwd=self.dirname)
+        else:
+            if not is_shallow:
+                subprocess.check_call(['git', 'clone', self.wrap.get('url'),
+                                       self.directory], cwd=self.subdir_root)
+                if revno.lower() != 'head':
+                    if subprocess.call(['git', 'checkout', revno], cwd=self.dirname) != 0:
+                        subprocess.check_call(['git', 'fetch', self.wrap.get('url'), revno], cwd=self.dirname)
+                        subprocess.check_call(['git', 'checkout', revno], cwd=self.dirname)
+            else:
+                subprocess.check_call(['git', 'clone', '--depth', self.wrap.values.get('depth'),
+                                       '--branch', revno,
+                                       self.wrap.get('url'),
+                                       self.directory], cwd=self.subdir_root)
+            if self.wrap.values.get('clone-recursive', '').lower() == 'true':
+                subprocess.check_call(['git', 'submodule', 'update',
+                                       '--init', '--checkout', '--recursive', '--depth', self.wrap.values.get('depth')],
+                                      cwd=self.dirname)
+            push_url = self.wrap.values.get('push-url')
+            if push_url:
+                subprocess.check_call(['git', 'remote', 'set-url',
+                                       '--push', 'origin', push_url],
+                                      cwd=self.dirname)
+
+    def is_git_full_commit_id(self, revno):
+        result = False
+        if len(revno) in (40, 64): # 40 for sha1, 64 for upcoming sha256
+            result = all((ch in '0123456789AaBbCcDdEeFf' for ch in revno))
+        return result
```

Comment:
Don't hardcode some homebrew heuristic to detect if something is a commit sha. It doesn't even matter whether something is a commit sha. What you actually want to know is if it is ineligible to be a clonable branch or tag.

For this, use `git ls-remote --heads --tags URL` and check if the output contains lines ending with `refs/heads/REVNO$` or `refs/tags/REVNO$`


```
$ git branch c762f19bc49b9c0552ff8c63c4637d9b85b6d26e
$ git log c762f19bc49b9c0552ff8c63c4637d9b85b6d26e
warning: refname 'c762f19bc49b9c0552ff8c63c4637d9b85b6d26e' is ambiguous.
Git normally never creates a ref that ends with 40 hex characters
because it will be ignored when you just specify 40-hex. These refs
may be created by mistake. For example,

  git switch -c $br $(git rev-parse ...)

where "$br" is somehow empty and a 40-hex ref is created. Please
examine these refs and maybe delete them. Turn this message off by
running "git config advice.objectNameWarning false"
fatal: bad object c762f19bc49b9c0552ff8c63c4637d9b85b6d26e
$ git log refs/heads/c762f19bc49b9c0552ff8c63c4637d9b85b6d26e
[content of logs for current HEAD appears]
```

Worse, you tried to be "future proof" with 64-hex ids, and git won't ever warn you about this, because 64-hex branch names is perfectly valid in current versions of git:

```
$ git branch c762f19bc49b9c0552ff8c63c4637d9b85b6d26efddfffdddfffdddfffdddfff
$ git log c762f19bc49b9c0552ff8c63c4637d9b85b6d26efddfffdddfffdddfffdddfff
[content of logs for current HEAD appears]
```

Reply:
I don't think i want to add complex code in a last minute regression fix like this. Also I don't think having branches or tags named like this is a real problem, but i'm happy to revisit this after the release to see what is the best solution.
Also what happens if the shape matching triggers will only break things if the git version is really old or if the build needs other tags or branches to be present.

Reply:
It's a good point that --exit-code is not portable to git less than 1.7.6 (you can simply omit that option and check for output lines ending in `refs/tags/REVNO$` or `refs/heads/REVNO$`).

Not sure what you mean by needing other tags or branches present, though.

Reply:
And yes, I acknowledge this is kind of an edge case, "it's unlikely" that people will have branch names that both have the exact number of characters and coincidentally only use hex characters, but I am a nitpicky person who nitpicks. :p

Reply:
Currently on master (without this PR) meson uses a manual emulation of git clone always. That emulation has 2 problems that i'm aware of: a) it does not work with centos7, b) it changes the behavior without the new `depth` option in a way that breaks `git describe`.

The way the logic is structured the git commit id string pattern matching is only used with `depth` so b) can not be triggered with this PR.

With `depth` the matching is used to switch to the manual clone emulation if `revno` looks like a commit id.

So the remaining broken case (as far as i know) is using branch or tag names that look exactly the same as commit ids (or future commit ids in the sha256 age) and using a very old git version. I think the best stability trade-of is to worry about this after the release if at all.